Implement smart hint system with advanced constraint analysis

hhyufan opened the pull request:
# Add Intelligent Hint System with Enhanced Algorithm

## 🎯 Overview
This PR introduces a comprehensive hint system that provides intelligent word suggestions to help players solve Wordle puzzles more strategically.

## ✨ Features Added

### 🧠 Smart Hint System
- **Intelligent word suggestions**: Advanced algorithm analyzes all previous guesses and feedback
- **Constraint-based filtering**: Considers confirmed positions, required letters, forbidden letters, and wrong positions
- **Duplicate letter handling**: Properly manages words with repeated letters through precise count constraints
- **Strategic scoring**: Ranks candidates based on letter frequency and information gain potential

### 🎯 Advanced Algorithm Features
- **Comprehensive constraint modeling**: Tracks all game state information from guess history
- **Smart red letter logic**: Distinguishes between completely absent letters vs wrong position letters
- **Multi-guess integration**: Analyzes entire game history collectively for optimal recommendations
- **Balanced selection**: Chooses from top-scoring candidates while maintaining variety

## 🎮 Usage
Players can call the hint system to receive intelligent word suggestions that:
- Respect all known constraints from previous guesses
- Maximize information gain for future guesses
- Handle complex scenarios including duplicate letters
- Provide strategic value beyond random suggestions

## 🔧 Technical Implementation
- Modular constraint system with separate building and validation phases
- Performance-optimized candidate filtering and scoring
- Robust error handling and edge case management
- Clean separation of concerns for maintainability

## 📈 Benefits
- Significantly improved hint accuracy compared to basic filtering
- Strategic assistance that helps players learn better Wordle techniques
- Enhanced user experience with more meaningful suggestions
- Solid foundation for future hint system enhancements

This implementation transforms the game from a pure puzzle into an educational tool that helps players understand optimal Wordle strategies.

hhyufan committed:
feat: add intelligent hint system for word suggestions

hhyufan committed:
feat: improve hint algorithm with comprehensive constraint analysis

- Redesign hint system with advanced constraint modeling
- Add intelligent word scoring based on letter frequency and information gain
- Implement precise letter count constraints for duplicate letter handling
- Enhance red letter logic to distinguish between absent vs wrong position
- Integrate all guess history for optimal word recommendations
- Replace random selection with smart candidate ranking system

TwAzuremy commented on main.py:
```diff
@@ -38,14 +38,13 @@ def parse_args():
             current_word = game.start(length)
+            ui.set_information("Start the game.")
+            ui.set_other_msg(f"{Fore.CYAN}[Command]{Fore.RESET} {Fore.RED}/exit{Fore.RESET} - exit the game, {Fore.YELLOW}/hint{Fore.RESET} - get a hint.")
+            ui.set_debug(f"The current word is {Fore.GREEN + current_word + Fore.RESET}")
```
An error may occur here before checking current_word.

TwAzuremy replied:
Please restore this first.

TwAzuremy replied:
Please restore this first.

hhyufan committed:
fix (Enter the length): Fixed an exception where the code execution order was wrong when the input exceeded the length of the range.

- Advance the verification of current_word correctness before setting the UI text information.

hhyufan committed:
delete: Remove incorrectly pushed image
